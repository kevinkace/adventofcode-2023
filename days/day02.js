
export function parseGames(games) {
    return games.split("\n").map((line) => {
        let [
            game,     // "Game 1"
            allRounds // "1 red, 1 green, 1 blue; 1 red, 1 blue; 1 blue"
        ] = line.split(": ");

        return {
            game   : parseInt(game.replace("Game ", ""), 10),

            rounds : allRounds.split("; ").map((round) => {
                return round.split(", ").reduce((moves, /** "1 green" */ move) => {
                    const [ count, color ] = move.split(" ");

                    moves[color.trim()] = parseInt(count, 10);

                    return moves;
                }, {});
            })
        };
    });
}