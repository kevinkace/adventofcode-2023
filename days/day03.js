export class Regions {
    constructor(str) {
        this.str    = str;
        this.parsed = str.split(/\r\n|\n/);

        this.regions  = [];
        this.specials = [];

        this.getRegions();
    }

    getRegions() {
        this.parsed.forEach((line, y) => {
            for (let x = 0; x < line.length; x++) {
                const currChar = line[x];

                if (/\d/.test(currChar)) {
                    this.addToRegion(currChar, x, y);

                    continue
                }

                this.closeRegion();

                if (currChar === ".") {
                    continue;
                }

                this.specials.push({ char : currChar, x, y });
            }

            this.closeRegion();
        });
    }

    addToRegion(number, x, y) {
        if (!this.currRegion) {
            this.currRegion = { number, x1 : x, y };

            return;
        }

        this.currRegion = {
            number : `${this.currRegion.number}${number}`,
            x1     : this.currRegion.x1,
            x2     : x,
            y      : y || this.currRegion.y
        }
    }

    closeRegion() {
        if (!this.currRegion) {
            return;
        }

        if (!this.currRegion.x2) {
            this.currRegion.x2 = this.currRegion.x1;
        }

        this.currRegion.number = parseInt(this.currRegion.number, 10);

        this.regions.push(this.currRegion);

        delete this.currRegion;
    }

    getCollisions() {
        return this.regions.filter((region, i) => {
            return this.specials.some((special, j) => {
                return specialCollision(region, special);
            });
        });
    }

    sumCollisions() {
        return this.getCollisions().reduce((acc, curr) => acc + curr.number, 0);
    }

}

export function specialCollision(region, special) {
    const x1 = region.x1 - 1;
    const x2 = region.x2 + 1;
    const y1 = region.y - 1;
    const y2 = region.y + 1;

    const collision = special.x >= x1 && special.x <= x2 && special.y >= y1 && special.y <= y2;

    if (!collision) {
        return false;
    }

    return true;
}

export function testCollision(box1, box2) {
    return box1.x1 < box2.x2 && box1.x2 > box2.x1 && box1.y1 < box2.y2 && box1.y2 > box2.y1;
}
