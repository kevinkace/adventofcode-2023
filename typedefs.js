/**
 * @typedef {Object} Game
 * @property {number} game
 * @property {Round[]} rounds
 */

/**
 * @typedef {Object} Round
 * @property {number} round
 * @property {number} red
 * @property {number} green
 * @property {number} blue
 */

export default {};
