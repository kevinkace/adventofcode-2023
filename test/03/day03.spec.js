import { readFile, writeFile } from "node:fs/promises";
import { fileURLToPath }       from "node:url";
import path                    from "node:path";

import { expect, test, describe } from "vitest";

import { parse } from "../../days/day02.js";

import typedefs from "../../typedefs.js";

import { Regions, sumIntegersAdjacentToSymbols, testCollision } from "../../days/day03.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const input    = await readFile(path.join(__dirname, "input.txt"), "utf8");
const sample01 = await readFile(path.join(__dirname, "sample01.txt"), "utf8");
const sample01Small = await readFile(path.join(__dirname, "sample01-small.txt"), "utf8");
const inputSmall = await readFile(path.join(__dirname, "input-small.txt"), "utf8");

import { samples } from "./samples.js";


describe("day 03", () => {
    // test("can parse input", () => {
    //     expect(parse(sample01)).toMatchSnapshot();
    // });

    test("can detect collisions", () => {
        expect(testCollision(
            { x1: 1, y1: 1, x2: 4, y2: 4 },
            { x1: 3, y1: 3, x2: 6, y2: 6 }
        )).toBe(true);
        expect(testCollision(
            { x1: 1, y1: 1, x2: 2, y2: 2 },
            { x1: 3, y1: 3, x2: 6, y2: 6 }
        )).toBe(false);
    });

    test("can get numbers", () => {
        const regions = new Regions(sample01);

        expect(regions).toMatchSnapshot();
    });

    test("gets collisions", () => {
        const regions = new Regions(sample01);

        expect(regions.getCollisions()).toMatchSnapshot();
    });

    test("answer pt 1", () => {
        let regions = new Regions(sample01);

        expect(regions.sumCollisions()).toEqual(4361);

        regions = new Regions(input);
        // const collisions = regions.getCollisions();
        const sum = regions.sumCollisions();
        // console.log("my sum: " + sum);
        // console.log(regions.regions.slice(80));
        expect(sum).toBeLessThan(3012388);
        expect(sum).toBeGreaterThan(43211);
        expect(sum).toBeGreaterThan(553499);
        expect(sum === 553585).toBe(false);
        expect(sum).toMatchSnapshot();
    });

    test.skip("med input", () => {
        let regions = new Regions(inputSmall);
        expect(regions.regions).toMatchSnapshot();
        expect(regions.specials).toMatchSnapshot();
        const collisions = regions.getCollisions();

        const sum = collisions.reduce((acc, curr) => {
            return acc + curr.number;
        }, 0);

        console.log(sum);
    });

    test.skip("samples", () => {
        samples.forEach(sample => {
            const regions = new Regions(sample);

            console.log(regions.sumCollisions());
        });
    });

    test("p2 sample", () => {
        const regions = new Regions(sample01);

        expect(regions.getGearRatios()).toMatchSnapshot();

        expect(regions.sumGearRatios()).toEqual(467835);
    });

    test("answer pt 2", () => {
        const regions = new Regions(input);

        expect(regions.sumGearRatios()).toMatchSnapshot();
    });
});
