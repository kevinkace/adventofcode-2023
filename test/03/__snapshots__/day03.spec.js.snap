// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 03 > answer pt 1 1`] = `554003`;

exports[`day 03 > answer pt 2 1`] = `87263515`;

exports[`day 03 > can get numbers 1`] = `
Regions {
  "parsed": [
    "467..114..",
    "...*......",
    "..35..633.",
    "......#...",
    "617*......",
    ".....+.58.",
    "..592.....",
    "......755.",
    "...$.*....",
    ".664.598..",
  ],
  "regions": [
    {
      "number": 467,
      "x1": 0,
      "x2": 2,
      "y": 0,
    },
    {
      "number": 114,
      "x1": 5,
      "x2": 7,
      "y": 0,
    },
    {
      "number": 35,
      "x1": 2,
      "x2": 3,
      "y": 2,
    },
    {
      "number": 633,
      "x1": 6,
      "x2": 8,
      "y": 2,
    },
    {
      "number": 617,
      "x1": 0,
      "x2": 2,
      "y": 4,
    },
    {
      "number": 58,
      "x1": 7,
      "x2": 8,
      "y": 5,
    },
    {
      "number": 592,
      "x1": 2,
      "x2": 4,
      "y": 6,
    },
    {
      "number": 755,
      "x1": 6,
      "x2": 8,
      "y": 7,
    },
    {
      "number": 664,
      "x1": 1,
      "x2": 3,
      "y": 9,
    },
    {
      "number": 598,
      "x1": 5,
      "x2": 7,
      "y": 9,
    },
  ],
  "specials": [
    {
      "char": "*",
      "x": 3,
      "y": 1,
    },
    {
      "char": "#",
      "x": 6,
      "y": 3,
    },
    {
      "char": "*",
      "x": 3,
      "y": 4,
    },
    {
      "char": "+",
      "x": 5,
      "y": 5,
    },
    {
      "char": "$",
      "x": 3,
      "y": 8,
    },
    {
      "char": "*",
      "x": 5,
      "y": 8,
    },
  ],
  "str": "467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..",
}
`;

exports[`day 03 > gets collisions 1`] = `
[
  {
    "number": 467,
    "x1": 0,
    "x2": 2,
    "y": 0,
  },
  {
    "number": 35,
    "x1": 2,
    "x2": 3,
    "y": 2,
  },
  {
    "number": 633,
    "x1": 6,
    "x2": 8,
    "y": 2,
  },
  {
    "number": 617,
    "x1": 0,
    "x2": 2,
    "y": 4,
  },
  {
    "number": 592,
    "x1": 2,
    "x2": 4,
    "y": 6,
  },
  {
    "number": 755,
    "x1": 6,
    "x2": 8,
    "y": 7,
  },
  {
    "number": 664,
    "x1": 1,
    "x2": 3,
    "y": 9,
  },
  {
    "number": 598,
    "x1": 5,
    "x2": 7,
    "y": 9,
  },
]
`;

exports[`day 03 > p2 sample 1`] = `
[
  [
    {
      "number": 467,
      "x1": 0,
      "x2": 2,
      "y": 0,
    },
    {
      "number": 35,
      "x1": 2,
      "x2": 3,
      "y": 2,
    },
  ],
  [
    {
      "number": 755,
      "x1": 6,
      "x2": 8,
      "y": 7,
    },
    {
      "number": 598,
      "x1": 5,
      "x2": 7,
      "y": 9,
    },
  ],
]
`;
